Buffer/scullBuffer.c

```diff
@@ -175,59 +172,63 @@ ssize_t scullBuffer_read(struct file *filp, char __user *buf, size_t count, loff
     up(&dev->sem);
     return 0;
   }
+
+  int currFront;
   currFront = dev->front;
+
   /* Move the front of the buffer up */
   dev->front += 516;
+  printk(KERN_DEBUG "scullBuffer: new front: %d\n", dev->front);
+
   up(&dev->sem);
 
-  /* Get mutex for accessing buffer */
+  /* Get semaphore for reading form buffer */
   if (down_interruptible(&dev->full))
     return -ERESTARTSYS;
 
   printk(KERN_DEBUG "scullBuffer: read called count= %d\n", count);
   printk(KERN_DEBUG "scullBuffer: front = %d \n", currFront);
 
-  /* Read a max of 512 bytes */
-  if( count > 512)
-    count = 512;
 
   /* Get mutex for accessing buffer */
   if (down_interruptible(&dev->sem))
     return -ERESTARTSYS;
 
   /* Read the size of the block first */
+  int res;
   res = kstrtoint(dev->bufferPtr + (currFront % (scull_size*516)), 10, &len);
   printk(KERN_DEBUG "scullBuffer: Item length is %d\n", len);
   if(res < 0) {
     up(&dev->sem);
+    up(&dev->empty);
     return -EINVAL;
   }
 
+  /* Read a max of len bytes */
+  if(len > 512) len = 512;
+  if(count > len) count = len;
+
   printk(KERN_DEBUG "scullBuffer: reading %d bytes\n", (int)count);
   /* copy data to user space buffer */
-  if (copy_to_user(buf, dev->bufferPtr + (currFront % (scull_size*516)) + 4, len)) {
-    countRead = -EFAULT;
+  if (copy_to_user(buf, dev->bufferPtr + (currFront % (scull_size*516)) + 4, count)) {
     up(&dev->sem);
-    return 0;
+    up(&dev->empty);
+    return -EFAULT;
```

Comment:
I don't know why returning 0 here, EFAULT seemed to make more sense,
and we should restore the semaphore because we should consider that item "read", we can't get it back.